feat(aio): report application errors to Google Analytics

petebacondarwin opened the pull request:
This is a basic implementation of error logging using the limited
facilities provided by Google Analytics.

Errors within the Angular app itself will be handled by a new
`ReportingErrorHandler` service, which overrides and extends the
built-in `ErrorHandler`.

Further, errors outside the app, which arrive at `window.onerror`
will also be reported to Google Analytics.

Closes #21943

petebacondarwin commented on aio/src/app/shared/ga.service.ts:
```diff
@@ -29,7 +29,15 @@ export class GaService {
+    const gaFn = (this.window as any)['ga'];
+    if (gaFn) {
+      gaFn(...args);
+    }
```
This guard was needed because errors in unit tests are now being posted to this service, even when the `window.ga` has not been set.

mary-poppins commented:
You can preview 46b4262 at https://pr22011-46b4262.ngbuilds.io/.

IgorMinar commented on aio/src/app/shared/ga.service.ts:
```diff
@@ -29,7 +29,15 @@ export class GaService {
 
+  sendError(error: any) {
+    const message = (error && error.message) || 'Unknown error thrown';
+    this.ga('send', 'exception', message.substr(0, 150));
```
it would be useful to have at least part of the stack. but of source that's very verbose, so how about we clean it up a bit with something like:

```js
message = e.stack
          // strip the leading "Error: "
          .replace(/^Error: /,'')
          // strip all leading "     at " for each frame
          .replace(/^ +at /gm, '')
          // replace long urls with just the last segment
          // this contains the filename, line+column number
          .replace(/ \(http.+\/(.+)\)/g, '@$1')
          // cap at 150 because that's the GA limit
          .substr(0, 150);
```

This is what it would look like: https://stackblitz.com/edit/angular-axqlkx?file=app%2Fapp.component.ts

Stack frames format vary from browser to browser, so we'd need to test this on various browsers to ensure that it does the right thing in most cases.

petebacondarwin replied:
OK, we can give that a go. As you say different browsers work differently so I was going with the simplest MVP to get us up and running.

IgorMinar commented on aio/src/app/shared/reporting-error-handler.ts:
```diff
@@ -0,0 +1,24 @@
+ * Errors outside the Angular application may also be handled by `window.onerror`.
+ */
+@Injectable()
+export class ReportingErrorHandler extends ErrorHandler {
```
why extend? you should really just implement the classes public interface. most of our classes are not intended to be extended like this.

petebacondarwin replied:
I extended because the default error handler does a great job of writing the error to the console, which IMO is very important. What it does is not trivial (see https://github.com/angular/angular/blob/master/packages/core/src/error_handler.ts#L45-L59) and I didn't want to have to reimplement that myself. I don't see any downside of extending rather than implementing in this case.

IgorMinar replied:
could we delegate to the parent instead? Maybe I'm being paranoid, but I always prefer composition/decoration over inheritance.

petebacondarwin replied:
How would we delegate? We would have to create our own instance of the `ErrorHandler` service, which is not much better than just extending it, since we would need to know how to create it.

What is the danger here? This is clearly an is-a relationship, and not an arbitrary reuse scenario.

IgorMinar replied:
I think you are right in this case. The danger is that we no longer ca add more arguments to the super method without a breaking change, but that would be true even if we were implementing the ErrorHandler as an interface.

IgorMinar commented on aio/src/app/shared/reporting-error-handler.ts:
```diff
@@ -0,0 +1,24 @@
+   * @param error Information about the error.
+   */
+  handleError(error: any) {
+    super.handleError(error);
```
if this handler throws we'll never get to send the error to GA

petebacondarwin replied:
I would hope that this was _very_ unlikely otherwise I would suggest that there is a bug in the default `ErrorHandler`. But we can wrap in a try-finally for safety.

IgorMinar replied:
For some reason I thought there were cases when the error handler would reformat and rethrow the error, but I don't think that's the case.

But now I'm wondering: let's say that the error is thrown from the handler, wouldn't the global handler catch it and report it?

IgorMinar replied:
oh, but it would report the error handling error and not the original error. but I guess that's ok.

IgorMinar commented on aio/src/index.html:
```diff
@@ -44,6 +44,13 @@
+  <script>
+    // Handle fatal errors that are outside Angular
+    window.onerror = function(message) {
+      ga('send', 'exception', {exDescription: message.substr(0, 150), exFatal: true});
```
this should use the same formatting as the service

petebacondarwin replied:
Sadly we can't ensure the same formatting because many browsers don't provide the error object (and so the stack trace) to this handler. But I guess for those that do we can try our best.

IgorMinar replied:
yeah. we should try our best. otherwise we'll have unactionable data.

IgorMinar reviewed (changes requested):
I'm very excited about this feature!!

IgorMinar commented:
results of a quick test on various browsers visiting https://angular-axqlkx.stackblitz.io

https://oasis.sandstorm.io/shared/E4xFr6rMYd5oqtzeVEMKtG__5cXFUlKt2vT_slvAzcW

Firefox on Android
```
shortened error message (length: 150):

AppComponent@https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29
createClass@https://angular-axqlkx.stackblitz.io/turbo_modul

original error message (length: 3134):

AppComponent@https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29
createClass@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12200:20
createDirectiveInstance@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12049:37
createViewNodes@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13487:53
createRootView@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13377:5
callWithDebugContext@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14778:39
debugCreateRootView@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14079:12
ComponentFactory_.prototype.create@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:10998:37
ComponentFactoryBoundToModule.prototype.create@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:3958:16
ApplicationRef.prototype.bootstrap@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5769:40
PlatformRef.prototype._moduleDoBootstrap/<@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5496:74
PlatformRef.prototype._moduleDoBootstrap@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5496:13
PlatformRef.prototype.bootstrapModuleFactory/</</<@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5417:21
ZoneDelegate.prototype.invoke@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:392:17
forkInnerZoneWithAngularBehavior/zone._inner<.onInvoke@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:4665:24
ZoneDelegate.prototype.invoke@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:391:17
Zone.prototype.run@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:142:24
scheduleResolveOrReject/<@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:873:52
ZoneDelegate.prototype.invokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:425:17
forkInnerZoneWithAngularBehavior/zone._inner<.onInvokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:4656:24
ZoneDelegate.prototype.invokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:424:17
Zone.prototype.runTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:192:28
drainMicroTaskQueue@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:602:25
[608]/E/</<@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:35475
[608]/E/<@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:35345
[564]/t.exports/a@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:26331
```



Safari 11
```
shortened error message (length: 150): 
    
AppComponent
createClass
createDirectiveInstance
createViewNodes
createRootView
callWithDebugContext
create
bootstrap
forEach@[native code]
_moduleDoB

original error message (length: 247):
    
AppComponent
createClass
createDirectiveInstance
createViewNodes
createRootView
callWithDebugContext
create
bootstrap
forEach@[native code]
_moduleDoBootstrap

onInvoke
run

onInvokeTask
runTask
drainMicroTaskQueue
promiseReactionJob@[native code]
```

Opera 50
```
shortened error message (length: 150):

something terrible has happened. oh no. oh no.
new AppComponent@app.component.ts:31:29
createClass@core.umd.js:12200:20
createDirectiveInstance@core.u

original error message (length: 1356):

Error: something terrible has happened. oh no. oh no.
    at new AppComponent (https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29)
    at createClass (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12200:20)
    at createDirectiveInstance (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12049:37)
    at createViewNodes (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13487:53)
    at createRootView (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13377:5)
    at callWithDebugContext (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14778:42)
    at Object.debugCreateRootView [as createRootView] (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14079:12)
    at ComponentFactory_.create (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:10998:46)
    at ComponentFactoryBoundToModule.create (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:3958:29)
    at ApplicationRef.bootstrap (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5769:57)
```

Chrome 64 (desktop + android)
```
shortened error message (length: 150):

something terrible has happened. oh no. oh no.
new AppComponent@app.component.ts:31:29
createClass@core.umd.js:12200:20
createDirectiveInstance@core.u

original error message (length: 1356):
---
Error: something terrible has happened. oh no. oh no.
    at new AppComponent (https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29)
    at createClass (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12200:20)
    at createDirectiveInstance (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12049:37)
    at createViewNodes (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13487:53)
    at createRootView (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13377:5)
    at callWithDebugContext (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14778:42)
    at Object.debugCreateRootView [as createRootView] (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14079:12)
    at ComponentFactory_.create (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:10998:46)
    at ComponentFactoryBoundToModule.create (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:3958:29)
    at ApplicationRef.bootstrap (https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5769:57)
```

Firefox 58
```
shortened error message (length: 150): 

AppComponent@https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29
createClass@https://angular-axqlkx.stackblitz.io/turbo_modul

original error message (length: 3014): 

AppComponent@https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/app.component.ts:31:29
createClass@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12200:20
createDirectiveInstance@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:12049:37
createViewNodes@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13487:53
createRootView@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:13377:5
callWithDebugContext@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14778:39
debugCreateRootView@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:14079:12
ComponentFactory_.prototype.create@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:10998:37
ComponentFactoryBoundToModule.prototype.create@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:3958:16
ApplicationRef.prototype.bootstrap@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5769:40
PlatformRef.prototype._moduleDoBootstrap/<@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5496:74
PlatformRef.prototype._moduleDoBootstrap@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5496:13
PlatformRef.prototype.bootstrapModuleFactory/</</<@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:5417:21
ZoneDelegate.prototype.invoke@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:392:17
onInvoke@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:4665:24
ZoneDelegate.prototype.invoke@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:391:17
Zone.prototype.run@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:142:24
scheduleResolveOrReject/<@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:873:52
ZoneDelegate.prototype.invokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:425:17
onInvokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/@angular/core@5.0.0/bundles/core.umd.js:4656:24
ZoneDelegate.prototype.invokeTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:424:17
Zone.prototype.runTask@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:192:28
drainMicroTaskQueue@https://angular-axqlkx.stackblitz.io/turbo_modules/zone.js@0.8.18/dist/zone.js:602:25
E/</<@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:35475
E/<@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:35345
a@https://c.staticblitz.com/assets/packs/common-7c635cf1d1bf1b16dfee.js:1:26331
```

IE 11
stackblitz doesn't have IE11 polyfills :-(


Edge 16
```
shortened error message (length: 150): 

something terrible has happened. oh no. oh no.
AppComponent (eval code:31:21)
createClass (eval code:12200:13)
createDirectiveInstance (eval code:1204

original error message (length: 549): 

Error: something terrible has happened. oh no. oh no.
   at AppComponent (eval code:31:21)
   at createClass (eval code:12200:13)
   at createDirectiveInstance (eval code:12049:5)
   at createViewNodes (eval code:13487:21)
   at createRootView (eval code:13377:5)
   at callWithDebugContext (eval code:14778:9)
   at debugCreateRootView (eval code:14079:5)
   at ComponentFactory_.prototype.create (eval code:10998:9)
   at ComponentFactoryBoundToModule.prototype.create (eval code:3958:9)
   at ApplicationRef.prototype.bootstrap (eval code:5769:9)
```

petebacondarwin commented:
So we should try to strip out the `@https://angular-axqlkx.stackblitz.io/tmp/appfiles/app/` part of the filename too.

petebacondarwin commented:
Here is a slightly refined trimming function that accounts for the Firefox format: https://stackblitz.com/edit/angular-kq6epp?file=app/app.component.ts

```
  extractMessage(e: Error) {
    // strip the leading "Error: " from the message
    const trimmedMessage = e.message.replace(/^Error: /,'');
    const trimmedStack = e.stack
          // strip the leading "Error: " from the stack trace
          .replace(/^Error: /,'')
          // strip the message from the stack trace
          .replace(trimmedMessage + '\n', '')
          // strip all leading "     at " for each frame
          .replace(/^ +at /gm, '')
          // replace long urls with just the last segment
          // this contains the filename, line+column number
          .replace(/(?: \(|@)http.+\/([^/]+)\)?\n/gm, '@$1\n');
    // cap at 150 because that's the GA limit
    return `${trimmedMessage}\n${trimmedStack}`.substr(0, 150);
  }
```

petebacondarwin commented on aio/src/app/shared/reporting-error-handler.ts:
```diff
@@ -0,0 +1,24 @@
+import { GaService } from './ga.service';
+
+/**
+ * Extend the default error handing to report errors to an external service - e.g Google Analytics.
```
handing -> handling

IgorMinar commented:
I like that new version of the transform, seems to do the right thing on all browsers we care about.

gkalpak commented on aio/src/app/shared/ga.service.ts:
```diff
@@ -29,7 +29,15 @@ export class GaService {
   }
 
+  sendError(error: any) {
+    const message = (error && error.message) || 'Unknown error thrown';
```
Would it make sense to use `error` if it is a string?

petebacondarwin commented:
After a fruitful discussion with @gkalpak I am going to put the trimming code inside the `onerror` handler and then call that from the `ReportingErrorHandler`.  This is the trimming code I am working with: https://gist.github.com/petebacondarwin/3a980ed8a7b74299ddb7381b74515614

mary-poppins commented:
You can preview de58ec9 at https://pr22011-de58ec9.ngbuilds.io/.

mary-poppins commented:
You can preview 9f2a4d5 at https://pr22011-9f2a4d5.ngbuilds.io/.

petebacondarwin committed:
feat(aio): report `logger.error` calls to Google Analytics

We have a number of observables that have `catch` handlers to recover
from errors without causing the stream to close, and breaking the app.
We also have some `try ... catch` blocks for synchronous code for a
similar reason.

In these cases we conventionally then call `logger.error` in the catch
handler. We are interested in these errors so we are going to capture them
by reporting them to Google Analytics via the new `ReportingErrorHandler`.

mary-poppins commented:
You can preview 6877641 at https://pr22011-6877641.ngbuilds.io/.

petebacondarwin commented on aio/src/app/shared/logger.service.ts:
```diff
@@ -1,18 +1,21 @@
   error(value: any, ...rest: any[]) {
-    console.error(value, ...rest);
+    const message = [value, ...rest].join(' ');
+    this.errorHandler.handleError(message);
```
If we are logging an error, then it probably needs to be reported to GA.

petebacondarwin replied:
This is the case for the cases where we use the `catch` RxJS operator to prevent streams from closing and killing the app when there is a recoverable error.

petebacondarwin commented on aio/src/app/shared/reporting-error-handler.ts:
```diff
@@ -0,0 +1,37 @@
+      this.reportError(e);
+    }
+    this.reportError(error);
+  }
```
We now catch exceptions thrown in the super class and report them too!

petebacondarwin commented on aio/src/index.html:
```diff
@@ -44,6 +44,38 @@
+          stack = url + ':' + line + ':' + col;
+        }
+        return (msg + '\n' + stack).substr(0, 150);
+      }
```
The error formatting (for GA) is kept in this one place (for DRYness), and the `ReportingErrorHandler` delegates to `onerror` for reporting the error to GA.

mary-poppins commented:
You can preview 3ba2494 at https://pr22011-3ba2494.ngbuilds.io/.

mary-poppins commented:
You can preview 9c0024d at https://pr22011-9c0024d.ngbuilds.io/.

gkalpak commented on aio/tests/e2e/onerror.e2e-spec.ts:
```diff
@@ -0,0 +1,202 @@
+      'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz' +
+      'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz',
+      'some-file.js', 12, 3, undefined);
+    expect(message).toEqual(  'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz' +
```
Whitespace after `(`.

petebacondarwin replied:
:+1:

petebacondarwin commented on aio/src/index.html:
```diff
@@ -44,6 +44,38 @@
+        if (e) {
+          stack = e.stack
+              // strip the leading "Error: " from the stack trace
+              .replace(/^Error: /,'')
```
:+1:

petebacondarwin commented on aio/src/index.html:
```diff
@@ -44,6 +44,38 @@
+              // strip all leading "at " for each frame
+              .replace(/^at /gm, '')
+              // replace long urls with just the last segment: `filename:line:column`
+              .replace(/(?: \(|@)http.+\/([^/)]+)\)?\n/gm, '@$1\n')
```
I yhink you are right and it makes more sense to be lenient here

petebacondarwin committed:
feat(aio): report application errors to Google Analytics

This is a basic implementation of error logging using the limited
facilities provided by Google Analytics.

Errors within the Angular app itself will be handled by a new
`ReportingErrorHandler` service, which overrides and extends the
built-in `ErrorHandler`.

Further, errors outside the app, which arrive at `window.onerror`
will also be reported to Google Analytics.

Closes #21943

mary-poppins commented:
You can preview b30c1b1 at https://pr22011-b30c1b1.ngbuilds.io/.
